### PR TITLE
Cs/sc 1389 exponential backoff

### PR DESCRIPTION
**Pull request by Charl1996**
I just created a draft so you can comment if you want before I merge into base branch.

### CODE REVIEW

File: corehq/motech/requests.py [hunk 1]
```diff
@@ -26,9 +31,23 @@
...
+SECONDS_IN_ONE_HOUR = 3600
+SECONDS_IN_ONE_DAY = 86400
 
-def log_request(self, func, logger):
+SECONDS_IN_PERIOD = {
+    'seconds': SECONDS_IN_ONE_SECOND,
+    'minutes': SECONDS_IN_ONE_MINUTE,
+    'hours': SECONDS_IN_ONE_HOUR,
+    'days': SECONDS_IN_ONE_DAY
+}
```
Comment (Charl1996) [inline corehq/motech/requests.py:47]: These are the defaults necessary for generic request backoff function (to be used as decorator)

File: corehq/motech/requests.py [hunk 2]
```diff
@@ -292,3 +311,42 @@ def validate_user_input_url_for_repeaters(url, domain, src):
                 'reason': e.reason,
             })
             raise
+
+
+def apply_request_backoff(
```
Comment (Charl1996) [inline corehq/motech/requests.py:316]: I decided to implement a kind of generic way to apply a backoff to functions making requests, hence implemented it as decorator for ease of use.

File: corehq/motech/requests.py [hunk 3]
```diff
@@ -292,3 +311,42 @@ def validate_user_input_url_for_repeaters(url, domain, src):
...
+    max_retries=MAX_TRIES,
+    exponential_base=DEFAULT_BACKOFF_RETRY_EXPONENT_BASE,
+    exponential_period=DEFAULT_EXPONENTIAL_PERIOD
```
Comment (Charl1996) [inline corehq/motech/requests.py:319]: Please feel free to comment on the naming of these two params; in my head it makes sense, but I'm not sure if it's immediately obvious to someone without context what they are and their use.

File: custom/onse/tasks.py [hunk 4]
```diff
@@ -208,6 +209,11 @@ def five_years_ago():
     return datetime.utcnow().date() - relativedelta(years=5)
 
 
+@apply_request_backoff(
+    max_retries=MAX_RETRIES,
+    exponential_base=EXPONENTIAL_BASE,
+    exponential_period=EXPONENTIAL_PERIOD
+)
 def fetch_data_set(
```
Comment (Charl1996) [inline custom/onse/tasks.py:217]: Are there other important places requests made to DHIS that also needs to be decorated that I might have missed?

File: corehq/motech/requests.py [hunk 5]
```diff
@@ -292,3 +311,42 @@ def validate_user_input_url_for_repeaters(url, domain, src):
...
+            retry_count = 0
+            response = None
+
+            while True:
+                try:
+                    response = request_function(*args)
+                except (RequestException, HTTPError):
+                    if retry_count < max_retries:
+                        backoff_period_seconds = _get_exponential_backoff_period_in_seconds(retry_count)
+                        sleep(backoff_period_seconds)
```
Comment (Charl1996) [inline corehq/motech/requests.py:332]: Are there better ways than to use the sleep() function? This is running in it's own thread so I don't foresee any issues, but just wondering.

**Review by snopoke — COMMENTED** [review]
Could we not use https://pypi.org/project/backoff/ instead of implementing it ourselves?

**Comment by Charl1996** [comment]
> Could we not use https://pypi.org/project/backoff/ instead of implementing it ourselves?

I guess that is an option. The package is handling lots of configurables, which is nice. I'll investigate a bit and see how easy it can be implemented. By the looks of it, it shouldn't be hard.

**Comment by Charl1996** [comment]
Closing this; going to use [backoff package](https://pypi.org/project/backoff/) instead.

**Pull request closed without merging**